### PR TITLE
Add textX test (for JSON right now)

### PR DESCRIPTION
**Pull request by ThatXliner**

### CODE REVIEW

File: bench/textx/json.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+from textx import metamodel_from_file
+from textx.export import metamodel_export, model_export
```
Comment (goodmami) [inline bench/textx/json.py]: These imports appear unused.

File: bench/textx/json.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+from os.path import join, dirname
+
+# STRING = _(r'"([ !#-\[\]-\U0010ffff]+|\\(["\/\\bfnrt]|u[0-9A-Fa-f]{4}))*"')
+# FLOAT = _(r"-?(0|[1-9][0-9]*)(\.[0-9]+)?([Ee][+-]?[0-9]+)?")
+# BOOL = _(r"(true|false)\b", rule_name="BOOL", root=True)
```
Comment (goodmami) [inline bench/textx/json.py]: These comments are unnecessary.

File: bench/textx/json.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+json_mm = metamodel_from_file(
+    join(this_folder, "json.tx"),
+    debug=False,
+    builtins={"FLOAT": float, "BOOL": bool, "STRING": str},
+    classes=[Member, Object],
+)
+
+
+def parse(s):
+    return json_mm.model_from_string(s)
```
Comment (goodmami) [inline bench/textx/json.py]: Should this be `json_mm.model_from_str(s)`?

**Review by goodmami — CHANGES_REQUESTED** [review]
Thanks!

I have a couple minor comments below, but also I'm getting an error:

```
    def parse(s):
>       return json_mm.model_from_string(s)
E       AttributeError: 'TextXMetaModel' object has no attribute 'model_from_string'
```

I think that call should use `model_from_str` instead.

Other things:

* If you wrap all the class definitions and metamodel creation in a function called `compile` and have it return the following:

  ```python
      return lambda s: json_mm.model_from_str(s)
  ```

  ... and redefine `parse` as:

  ```python
  parse = compile()
  ```

  ...then the benchmarks can test the time it takes to instantiate a parser, as well (separately from parsing time).

* Even with the changes outlined above, I'm getting validation errors:
  - The result of parsing `null` is a string `"null"` and not `None`
  - Objects and Arrays both return textX model classes and not Python dicts and lists, respectively.
  - Strings are not unescaped. You can use `bench.helpers.json_unescape()` for this.

The tests stop validating after the first error in each test function, so I'm not sure if there are other errors. Try running `python validate.py` to test things out.

Could you please fix the above issues and push a new commit?

**Comment by ThatXliner** [comment]
Yeah sure! This is just the initial implementation I hacked up (the documentation really sucks).

**Comment by goodmami** [comment]
> (the documentation really sucks)

Agreed. After 20 minutes of trying to fix the issues my self I gave up. I just wanted to know how to apply semantic actions to parsing events. I think it has to do with registering interpreter functions, but would it be model interpreters or object interpreters? Anyway, good luck!

**Comment by ThatXliner** [comment]
That was probably the 5th hackiest code I've ever written. But I'm pretty there is no other way than resort to `collections.UserDict` and `collections.UserList`

**Comment by ThatXliner** [comment]
What's the difference between `*` and `*-compile` (e.g. `JSON` and `JSON-compile`)? Lark is the slowest for `JSON-compile` but the 3rd fastest for `JSON`.

Also, what is the data you use to shove into term-graph? The median? The mean? It depends (on if it's skewed)?

**Comment by goodmami** [comment]
Some answers to your questions before I review the PR:

> What's the difference between `*` and `*-compile` (e.g. `JSON` and `JSON-compile`)? Lark is the slowest for `JSON-compile` but the 3rd fastest for `JSON`.

`*-compile` is the time it takes to initialize the parser, which is why all the setup code goes in a `compile()` function. Tools that initialize the parser by parsing a grammar description (pe, Lark, Parsimonious) tend to be slower and those that are created with Python code (Sly, PyParsing) tend to be faster. It has no bearing on the speed of parsing, but a holistic view of a program that parses things might consider the initialization time.

> 
> Also, what is the data you use to shove into term-graph? The median? The mean? It depends (on if it's skewed)?

I believe I used the mean as I was trying to avoid outliers. If the deviation was too high, I'd re-run the benchmarks with fewer things running in the background. I also generally close the browser and anything that is streaming, like music, etc., before running the benchmarks. But don't worry about adding the time to the graph. I'll do that after rerunning everything. They need to be run on the same machine.

**Comment by ThatXliner** [comment]
Alrighty, that answers my questions. But just one more thing.

As you see with my last commit: 
Screen Shot 2021-04-19 at 6 21 34 PM
I decided to use [packrat parsing](http://textx.github.io/textX/stable/parser_config/#memoization-aka-packrat-parsing) which would use more memory (packrat parsing is just, AFIAK, a fancy name for PEG parsing with a [memoization cache](https://stackoverflow.com/q/6469437/15396573))

Is memory usage a thing we're also testing for?

File: bench/textx/json.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        def __iter__(self):
+            return iter(self.pair)
+
+    class Object(collections.UserDict):
+        def __init__(self, members, parent=None):
+            self.data = {key: value for key, value in members}
+
+    class Array(collections.UserList):
+        def __init__(self, values, parent=None):
+            self.data = list(values)
```
Comment (goodmami) [inline bench/textx/json.py]: It's not actually necessary to use the UserDict/UserList classes. The following works, too:

```python
    class Object(dict):
        def __init__(self, members, parent=None):
            self.update((m.key, m.value) for m in members)

    class Array(list):
        def __init__(self, values, parent=None):
            self.extend(values)
```

And note that there's no need for importing `collections` and the `Members` class can be completely removed.

File: bench/textx/json.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    json_mm = metamodel_from_file(
+        join(dirname(__file__), "json.tx"),
+        debug=False,
+        classes=[Member, Object, Array],
+        memoization=True,
```
Comment (goodmami) [inline bench/textx/json.py]: For JSON, memoization actually hurts performance because there's basically no backtracking needed.

**Review by goodmami — APPROVED** [review]
This is much better, and now it passes validation. I'll approve the PR but you may want to make some modifications as described below.

The textX benchmark is very slow and memory-hungry. I tried parsing the "big" JSON string as in the tests just once, and here's the time/memory readings I get (note: these are command-line timings and therefore include initialization time):

| Benchmark | Time | Memory |
| --- | --- | --- |
| stdlib | 0.4s | 46M |
| pe | 2.0s | 54M |
| sly | 4.2s | 54M |
| lark | 4.2s | 64M |
| parsimonious | 9.7s | 600M |
| pyparsing | 12.2s | 61M |
| textx | 16.7s | 390M |
| textx' | 14.7s | 335M |

The last line (`textx'`) is after I made some modifications (see comments in code), which I was able to discover thanks to your work figuring out how to use textX:
- Disable memoization (-1.8s)
- Simplify classes (-0.2s)

You might try asking the textX community for other performance improvements as it's even slower than pyparsing (the previous slowest by a wide margin), and the second most memory-hungry, even with memoization turned off.

*edit: added sly benchmark*

File: bench/textx/json.py (hunk 4 above)
Comment (goodmami) [reply bench/textx/json.py]: To clarify: You can remove the `collections` import if you use the simplified classes. And there was no speed improvement to using these classes over the `User*` ones, but the 0.2s was gained by removing the custom `Member` class and just using the default textX one in the `Object.__init__()` method.

**Comment by ThatXliner** [comment]
I think I can merge it for now

**Comment by ThatXliner** [comment]
I think that's all we need. Ready for merge

**Comment by goodmami** [comment]
Yep looks good. I've squashed all the commits into one and merged. There's a few small things to fix, but I'll take care of it. Thanks!

**Comment by ThatXliner** [comment]
You're welcome! Happy to help 😄

**Pull request merged**